kubernetes/README.md

```diff
@@ -0,0 +1,107 @@
+
+- shopping (loopback4-example-shopping): an application developed with LoopBack
+  4, including APIs and implementations for user, order, and shopping-cart
+- recommender (loopback4-example-recommender): an Express application to mock up
+  the product recommendation service
+
+In addition, two resources are required for the application:
+
+- A MongoDB database for users and orders
+- A Redis registry for shopping cart items
+
+There are a few approaches that we use to run or test the application:
+
+1. Use `recommender` as a development dependency for `shopping` and invoke
+   recommender service in the same process.
+2. Use [concurrently](https://github.com/kimmobrunfeldt/concurrently) to start
+   `recommender` and `shopping` as two local processes. The communication
+   between `shopping` and `recommender` is over REST or gRPC.
+
+Similarly, we install/start `mongodb` and `redis` in different ways:
+
+1. Install `mongodb` and `redis` locally and start them as local processes
+2. Use `travis` services for `mongodb` and `redis` for the CI
+3. Use `docker` to start `mongodb` and `redis` containers.
+
+You may start to wonder what's the best practice to deploy a composite
+application like the shopping example that consists of multiple microservices,
+especially to cloud environments. To answer such questions, we did some
+experiment to explore how to bundle and deploy the shopping application as a
+Kubernetes cluster. It led to the introduction of Kubernetes based deployment to
+remove inconsistencies and promote cloud-native microservices.
+
+# Kubernetes based deployment
+
+The shopping example application consists of multiple microservices that are
+deployed as Docker containers managed by a kubernetes cluster.
+
+This `kubernetes` directory contains the scripts and helm chart for the shopping
+application.
+
+![shopping-app cluster](./imgs/k8s-shopping-cluster.png)
+
+It takes us a few steps:
+
+## Enable gRPC communication between shopping and recommender
+
+REST APIs are used to connect `shopping` to `recommender`. To better facilitate
+cloud-native deployment, we add the capability using `loopback-connector-grpc`
+to showcase the usage of [gRPC](https://grpc.io/), which is highly-efficient RPC
+protocol for communication between microservices.
+
+## Build docker images
+
+We leverage multi-stage build to create docker images for `shopping` and
+`recommender` microservices.
+
+- Stage 1: Build deployable packages using lerna
+
+  - Dockerfile.monorepo
+
+- State 2: Copy `shopping` and `recommender` packages into their own images
+  - Dockerfile.shopping
+  - Dockerfile.recommender
+
+```sh
+npm run docker:build
+```
+
+## Organize deployment as an Helm chart
+
+Instead of deploying each docker images by hand, we use `Helm` chart to describe
+the composition of the application using Kubernetes artifacts. The chart is
+described in `shopping-app`.
+
+As illustrated above, the chart includes two deployments and corresponding
```

Comment:
> As illustrated above, the chart includes two deployments and corresponding services 

Just want to be clear on terminology used in this sentence.

Aren't the recommender and shopping and redis and mongo considered services? (microservices).

Also, aren't all of them deployments to our cluster?

In our Helm chart we provide the helm info for shopping and recommender, but use a public helm chart for redis and mongodb (that other people defined)?

Perhaps the new sentence should be:

`...the chart deploys recommender, shopping, mongodb, and redis to the cluster` .